lib/posix/semaphore.c

```diff
@@ -1,11 +1,76 @@
 /*
  * Copyright (c) 2018 Intel Corporation
+ * Copyright (c) 2023 Meta
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
 #include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/posix/fcntl.h>
 #include <zephyr/posix/pthread.h>
+#include <zephyr/posix/semaphore.h>
+
+struct nsem_obj {
+	sys_snode_t snode;
+	sem_t sem;
+	atomic_t ref_count;
+	char *name;
+};
+
+/* Initialize the list */
+sys_slist_t sem_list = SYS_SLIST_STATIC_INIT(&sem_list);
+
+K_MUTEX_DEFINE(nsem_mutex);
+
+#if defined(__sparc__)
```

Comment:
Instead of directly depend on board, perhaps depend on a config that is only set for needed board.

Reply:
@cfriedt I simply copied this from `mqueue.c`, should we place this somewhere more generic?

Reply:
My comment is related to the fact that we imho should have checks depending on features and not directly boards. The board config can then enable them, or we could enable the features by default for certain boards.

Reply:
Definitely the malloc aspect is nice to have @ycsin.

I actually would really love to apply a simply pattern to all of the "pooled" types in posix, and we actually really also need this to separate pthreads / thread stacks / c11 threads, etc.

In any case though, solve that pattern once and then it can be used over and over again in posix, for sure.

What about this convention for mixing static and dynamic storage for "pooled types", @jukkar ?

* If the size is negative (Kconfig allows negative ints)
  - the absolute value of the size (minus one) is used for number of static allocations in the pool
  - the negative bit indicates that items can also be allocated from the heap
* If it's positive only, then that would mean only static allocations

Reply:
Fixed and rebased